central/cloudsources/manager/manager.go

```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+// Start the collection of assets from cloud sources.
+func (m *managerImpl) Start() {
+	m.loopTicker = time.NewTicker(m.loopInterval)
+	go m.discoveredClustersLoop()
+}
+
+// Stop the collection of assets from cloud sources.
+func (m *managerImpl) Stop() {
+	m.stopSignal.Signal()
+}
+
+// ShortCircuit the collection of assets from cloud sources.
+func (m *managerImpl) ShortCircuit() {
+	m.shortCircuitSignal.Signal()
+}
+
+func (m *managerImpl) discoveredClustersLoop() {
+	defer m.loopTicker.Stop()
+
+	for {
+		select {
+		case <-m.shortCircuitSignal.Done():
+			if err := m.shortCircuitRateLimiter.Wait(concurrency.AsContext(&m.stopSignal)); err != nil {
+				if errors.Is(err, context.Canceled) {
+					return
+				}
+				log.Errorw("Waiting for rate limiter entrance", logging.Err(err))
+			}
+			// Make sure to reset the signal again.
+			m.shortCircuitSignal.Reset()
+			// Make sure to reset the ticker, so we are not in the case where short-circuit is called and shortly after
+			// the interval is reached and discovered clusters are gathered again.
+			m.loopTicker.Reset(m.loopInterval)
+			m.discoverClustersFromCloudSources()
+		case <-m.loopTicker.C:
+			m.discoverClustersFromCloudSources()
+		case <-m.stopSignal.Done():
+			return
+		}
+	}
+}
+
+func (m *managerImpl) discoverClustersFromCloudSources() {
+	clusters := m.getDiscoveredClustersFromCloudSources()
+
+	m.reconcileDiscoveredClusters(clusters)
+}
+
+func (m *managerImpl) getDiscoveredClustersFromCloudSources() []*discoveredclusters.DiscoveredCluster {
+	// Fetch the cloud sources from the datastore. This will ensure that we will always use the latest updates.
+	// For the time being we do not foresee this to be a high cardinality store.
+	cloudSources, err := m.cloudSourcesDataStore.ListCloudSources(cloudSourceCtx, search.EmptyQuery())
+	if err != nil {
+		log.Errorw("Failed listing stored cloud sources", logging.Err(err))
+		return nil
+	}
+
+	clients := createClients(cloudSources)
+
+	var clientErrors error
+	var discoveredClusters []*discoveredclusters.DiscoveredCluster
+	for _, client := range clients {
+		resp, err := client.GetDiscoveredClusters(context.Background())
+		if err != nil {
+			clientErrors = errors.Join(clientErrors, err)
+			continue
+		}
+		discoveredClusters = append(discoveredClusters, resp...)
+	}
+
+	if clientErrors != nil {
+		log.Errorw("Received errors during fetching assets from cloud sources. The result might be incomplete",
+			logging.Err(clientErrors))
+	}
+	log.Debugf("Got the following discovered clusters from Cloud Source integrations: %+v", discoveredClusters)
+	return discoveredClusters
+}
+
+func (m *managerImpl) reconcileDiscoveredClusters(clusters []*discoveredclusters.DiscoveredCluster) {
+	// Remove discovered clusters which are out of the retention time.
+	deleted, err := m.discoveredClustersDataStore.DeleteDiscoveredClusters(discoveredClusterCtx, search.NewQueryBuilder().
```

Comment:
Can we re-use the centralized pruning mechanism we used for admin events here?

Reply:
I've re-written it a bit, and exposed the pruning function on the datastore instead of having the manager directly call the pruning function.